### ISSUE TITLE
PyPI Release

### ISSUE THREAD

**Issue by awsaf49** [sebastian-sz/resnet-rs-keras#12]
I think it would be amazing to have it on PyPI. Any plan to publish it there? @sebastian-sz

**Comment by sebastian-sz** [comment]
Hi @awsaf49  thank you for the kind words!

I actually thought about releasing on PyPI and decided **not to** do it because of two reasons:

#### Reason 1. Increased maintenance cost.
When creating this repo I aimed for it to be not only usable and easy to set up, but also for it to have as low maintenance cost as possible - ML models once released does not change too often (unless they are continuously developed).

For each new Tensorflow version I am testing the basic functionality. If all tests pass - I just merge to master. PyPI release will create an additional "layer" of maintenace, where I would have to create a release with each new Tensorflow version (or risk releasing broken package). Plus there is juggling with different Python / Tensorflow versions. 

Yes, it could probably be automated, but this would require some amount of work, that I currently do not have resources to do. Plus, it also takes us to Reason 2.

#### Reason 2. Not much benefit for the end user.
The models, in the current state, are still installable via `pip`. This works in terminal, or in `requirements.txt` or even in `Dockerfile`. The only thing that would change is that the end user would write, something like, `pip install resnet-rs-keras`, instead of `pip install git+(...)`

I think this change is too small to justify package release, outside of GitHub.

#### To wrap it up.
So, to summarize: I decided not to release on PyPI because it's additional work with not much gain (in my opinion).  
I hope this answers your question!

**Comment by awsaf49** [comment]
I'm not an expert on `PyPI` but I think there are some advantages over `pip install git+`,
* Installing different versions is a bit easier from **PyPI** as **PyPI** keeps track of your release and you can simply mention the version to install a specific version, `pip install resnet_rs==1.0`. I'm not sure if we can do the same thing from **GitHub**.
* There are someplace where you don't have access to internet(kaggle code competition), there we need to use **wheel** files `.whl` to install libraries and downloading them from **PyPI** is really easy, I don't know if `GitHub` also provide **wheel** files or not.

However, if you decide to publish it on **PyPI** or not, I would like to thank you for open-sourcing your project :)

**Comment by sebastian-sz** [comment]
### Versioning
Yes it is also possible with GitHub :) You can even install specific commit.  

For example in this repository:
* To install from branch v1.0 (before TPU fix): `pip install git+https://github.com/sebastian-sz/resnet-rs-keras@v1.0`
* To install from brach v1.1 (after TPU fix): `pip install git+https://github.com/sebastian-sz/resnet-rs-keras@v1.1` (at this moment this is the same as installing from main).
* If you want to go lower level and lock the version to a specific commit with SHA `0291dfdef96b63d1f66b9809c619f900fbb674ee` you could do:
`pip install git+https://github.com/sebastian-sz/resnet-rs-keras@0291dfdef96b63d1f66b9809c619f900fbb674ee`

It is advised to install from main. I almost never change the models, unless there are some serious functionalities broken or incompatiblities (like the mentioned TPU error). The majority of commits are internals, like updating test code or bump TF version - these do not change anything for the end user.

### Wheel files
In this case building wheel files can be done manually:

1. `git clone ...` && `cd resnet-rs-keras/`
2. `python setup.py bdist_wheel` 
3. `pip install dist/resnet_rs_keras-1.1-py3-none-any.whl`

**Comment by sebastian-sz** [comment]
I'm closing this issue, but feel free to reopen if you have any more questions!

**Comment by awsaf49** [comment]
> ### Versioning
> 
> Yes it is also possible with GitHub :) You can even install specific commit.
> 
> For example in this repository:
> 
>     * To install from branch v1.0 (before TPU fix): `pip install git+https://github.com/sebastian-sz/resnet-rs-keras@v1.0`
> 
>     * To install from brach v1.1 (after TPU fix): `pip install git+https://github.com/sebastian-sz/resnet-rs-keras@v1.1` (at this moment this is the same as installing from main).
> 
>     * If you want to go lower level and lock the version to a specific commit with SHA `0291dfdef96b63d1f66b9809c619f900fbb674ee` you could do:
>       `pip install git+https://github.com/sebastian-sz/resnet-rs-keras@0291dfdef96b63d1f66b9809c619f900fbb674ee`
> 
> 
> It is advised to install from main. I almost never change the models, unless there are some serious functionalities broken or incompatiblities (like the mentioned TPU error). The majority of commits are internals, like updating test code or bump TF version - these do not change anything for the end user.
> ### Wheel files
> 
> In this case building wheel files can be done manually:
> 
>     1. `git clone ...` && `cd resnet-rs-keras/`
> 
>     2. `python setup.py bdist_wheel`
> 
>     3. `pip install dist/resnet_rs_keras-1.1-py3-none-any.whl`

Thanks for your answer :)
